[ENG-3854] Add management command to sync all DOI metadata

Johnetordoff opened the pull request:
<!-- Before submit your Pull Request, make sure you picked
     the right branch:

     - For hotfixes, select "master" as the target branch
     - For new features, select "develop" as the target branch
     - For release feature fixes, select the relevant release branch (release/X.Y.Z) as the target branch -->

## Purpose

<!-- Describe the purpose of your changes -->

## Changes

<!-- Briefly describe or list your changes  -->

## QA Notes

Please make verification statements inspired by your code and what your code touches.
- Verify
- Verify

What are the areas of risk?

Any concerns/considerations/questions that development raised?

## Documentation

<!-- Does any internal or external documentation need to be updated?
     - If the API was versioned, update the developer.osf.io changelog.
     - If changes were made to the API, link the developer.osf.io PR here.
-->

## Side Effects

<!-- Any possible side effects? -->

## Ticket

https://openscience.atlassian.net/browse/ENG-3854

mfraezz commented on osf/management/commands/sync_doi_metadata.py:
```diff
@@ -0,0 +1,68 @@
+@app.task(name='osf.management.commands.sync_doi_metadata', max_retries=5, default_retry_delay=60)
+def sync_identifier_doi(identifier):
+    identifier.referent.request_identifier_update('doi')
+    identifier.modified = timezone.now()
```
Minor: `.save` will update this as it's `auto_now=True`; there's no real need to set it.

mfraezz commented on osf/management/commands/sync_doi_metadata.py:
```diff
@@ -0,0 +1,68 @@
+            help='number of dois to update in this batch.',
+        )
+        parser.add_argument(
+            '--modified_date',
```
Should this be `required=True`? It's a required positional arg for `sync_doi_metadata` with no default.

Johnetordoff commented on osf/management/commands/sync_doi_metadata.py:
```diff
@@ -0,0 +1,68 @@
+
+@app.task(name='osf.management.commands.sync_doi_metadata', max_retries=5, default_retry_delay=60)
+def sync_identifier_doi(identifier):
+    identifier.referent.request_identifier_update('doi')
```
@jwalz Thoughts on future-proofing this? is it worth it?

jwalz replied:
I'd add object_id__isnull=False to the identifiers query below. It's easy and makes this reusable

Johnetordoff commented:
@jwalz  @mfraezz I've addressed the comments, but one additional thing I had to do was move the test  under the `api_test` directory, otherwise the conftest.py file under `osf_tests` does not start the celery app for the GHA environment. IT was easier to move the test then to deal with potential test side effects from tampering with the `osf_test` conftest.py. Other then that should be g2g. 

Visibility for Matt these old EZID DOIs should be fixed by this: https://openscience.atlassian.net/browse/ENG-3933

jwalz commented:
> Visibility for Matt these old EZID DOIs should be fixed by this: https://openscience.atlassian.net/browse/ENG-3933

Do we have confirmation on this? Are we sure the EZID DOIs aren't the 'legacy_doi' dois?

Johnetordoff commented:
I assume as long as they Identifiers they will grab their refferent's corresponding client and updating the the DOI via that, so the Datacite client will delete the metadata for those, I believe.

jwalz commented on api_tests/identifiers/managment_commands/test_sync_dois.py:
```diff
@@ -0,0 +1,71 @@
+        assert registration_identifier.modified.date() < datetime.datetime.now().date()
+
+        call_command('sync_doi_metadata', f'-m={datetime.datetime.now()}')
+        assert len(mock_datacite.calls) == 2
```
nit: why 2?

Johnetordoff replied:
One call mints the DOI and the other uploads the metadata, the full details are in the client update logic.

jwalz commented on osf/management/commands/sync_doi_metadata.py:
```diff
@@ -0,0 +1,68 @@
+                f'{identifiers.count()} identifiers to mint')
+
+    for identifier in identifiers:
+        if not dry_run:
```
Let's make sure that the referent isn't deleted or withdrawn (or spam? or should deleted catch that?) before we spin up the task.

Johnetordoff replied:
The client sends a withdraw request when it has a deleted/private node, I was hoping that would update some of those old EZID DOIs, and potentially other's that might have randomly not been contacted to network issues.

mfraezz replied:
Yep. `not identifier.referent.is_retracted` should probably also be checked below, too.
